From 26d1dce50e012c75310056ebf3f0f08ff09b2820 Mon Sep 17 00:00:00 2001
From: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Date: Mon, 15 May 2023 13:18:57 +0300
Subject: [PATCH] drivers: xen: add Xen extended regions driver

Introduce a driver for the extended regions provided by Xen to Zephyr,
running as Xen Domain-0. Extended regions are the unmapped spaces in
Stage 2 address space, and can be used by Zephyr for foreign mapping,
i.e. mapping pages managed by Xen or belonging to other domains. The
regions are configured via the regs property in the hypervisor node,
compatible with xen,xen.

For example, the following device-tree overlay:

&hypervisor {
    reg = < 0x0 0x40200000 0x0 0x1000000
            0x0 0x40400000 0x0 0x1FC00000
            0x0 0x80000000 0x1 0x40000000>;
};

will allocate two extended regions:
- 0x40400000 with size 0x1fc00000
- 0x80000000 with size 0x140000000

The first region, 0x40200000 (size 0x1000000), is used for grant_tables
and will be ignored.

Backported to Zephyr v4.4.1: v4.4.1 was cut from the v4.4 maintenance branch
(forked at v4.4.0, 2026-04-14) and does not contain this commit.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Svitlana Drozd <svitlana_drozd@epam.com>
Signed-off-by: Yuichi Kusakabe <yuichi_kusakabe@jp.honda>
Upstream-Status: Backport [https://github.com/zephyrproject-rtos/zephyr/commit/26d1dce50e012c75310056ebf3f0f08ff09b2820]
---
diff --git a/drivers/xen/CMakeLists.txt b/drivers/xen/CMakeLists.txt
index d8d0d3caaa6..d25a7d9fc51 100644
--- a/drivers/xen/CMakeLists.txt
+++ b/drivers/xen/CMakeLists.txt
@@ -7,5 +7,6 @@ zephyr_sources_ifdef(CONFIG_XEN_GRANT_TABLE gnttab.c)
 zephyr_sources(memory.c)
 zephyr_sources_ifdef(CONFIG_XEN_DMOP dmop.c)
 zephyr_sources(version.c)
+zephyr_sources_ifdef(CONFIG_XEN_REGIONS regions.c)
 
 add_subdirectory_ifdef(CONFIG_XEN_DOM0 dom0)
diff --git a/drivers/xen/Kconfig b/drivers/xen/Kconfig
index 319f5b8a069..9ac0c8e20b8 100644
--- a/drivers/xen/Kconfig
+++ b/drivers/xen/Kconfig
@@ -1,5 +1,5 @@
 # SPDX-License-Identifier: Apache-2.0
-# Copyright (c) 2022-2024 EPAM Systems
+# Copyright (c) 2022-2026 EPAM Systems
 
 if XEN
 
@@ -11,9 +11,21 @@ config XEN_DMOP
 	  Enable lightweight wrappers for Xen dm_op hypercalls used by
 	  Xen device backends. Disable to drop dmop.c from the build.
 
+config XEN_REGIONS
+	bool "Xen extended regions driver"
+	depends on MMU
+	select SYS_MEM_BLOCKS
+	select KERNEL_DIRECT_MAP
+	help
+	  Xen extended regions driver. Allows using extended regions, provided by
+	  XEN hypervisor to Zephyr. This memory can be used to avoid allocations
+	  from Zephyr heap memory in favour of free unmapped space on Stage 2
+	  translation level. Requires the presence of at least one extended region
+	  in the hypervisor node.
+
 config XEN_GRANT_TABLE
 	bool "Xen grant table driver"
-	depends on HEAP_MEM_POOL_SIZE > 0
+	depends on HEAP_MEM_POOL_SIZE > 0 || XEN_REGIONS
 	default y
 	help
 	  Xen grant table driver. Please note that driver uses dynamic memory
diff --git a/drivers/xen/regions.c b/drivers/xen/regions.c
new file mode 100644
index 00000000000..68738f95d6f
--- /dev/null
+++ b/drivers/xen/regions.c
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) 2026 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/arch/arch_interface.h>
+#include <zephyr/sys/mem_blocks.h>
+#include <zephyr/xen/generic.h>
+
+LOG_MODULE_REGISTER(xen_regions);
+
+#define EXTENDED_REGIONS_START_IDX 1
+#define XEN_HYP_NODE DT_INST(0, xen_xen)
+
+static K_MUTEX_DEFINE(ext_regions_lock);
+
+struct extended_regions {
+	sys_mem_blocks_t *block;
+	uint64_t phys_addr;
+	uint64_t size;
+};
+
+/*
+ * At least one extended region should be provided in the hypervisor node.
+ * If there is no extended regions - alternative allocation should be implemented
+ * by the driver user and driver should be disabled.
+ */
+BUILD_ASSERT(DT_NUM_REGS(XEN_HYP_NODE) > EXTENDED_REGIONS_START_IDX,
+	     "No Extended regions were provided in DT");
+
+#define NOT_OVERLAPPING_BOUNDARIES(x_start, x_end, x_size, y_start, y_end, y_size) \
+	MAX(x_end, y_end) - MIN(x_start, y_start) >= x_size + y_size
+
+#define NOT_OVERLAPPING(x_start, x_size, y_start, y_size)		\
+	NOT_OVERLAPPING_BOUNDARIES((uint64_t)x_start,			\
+				  (uint64_t)x_start + (uint64_t)x_size, \
+				  (uint64_t)x_size,			\
+				  (uint64_t)y_start,			\
+				  (uint64_t)y_start + (uint64_t)y_size, \
+				  (uint64_t)y_size)
+
+/*
+ * Add BUILD_ASSERT that rejects extended region, starting from 0x0.
+ * This is done because mem_blocks is not designed to handle buffer
+ * starting from 0 and has assert with check. In this case starting address
+ * should be shifted to page 1. That shouldn't be a problem because
+ * minimal extended region size is set to 64MB in XEN.
+ * For example for the extended region 0->0x30000000 the following
+ * regs should be provided:
+ * regs = <.... 0x0 0x1000 0x0 0x2FFFF000 ... >;
+ *
+ * Check reg addr and reg size to be aligned to
+ * CONFIG_MMU_PAGE_SIZE. This is the requirement for arch_mem_map call.
+ *
+ * Extended regions should not overlap kernel Virtual Memory region.
+ */
+#define EXT_BLOCK(node_id, idx)						\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),		\
+		     "Extended region starting from 0 is not supported, please skip page 0"); \
+	BUILD_ASSERT(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region size is not aligned");		\
+	BUILD_ASSERT(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx) ==		\
+		     ROUND_DOWN(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx), CONFIG_MMU_PAGE_SIZE), \
+		     "Extended region address is not aligned");		\
+	BUILD_ASSERT(NOT_OVERLAPPING(DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx), \
+				     DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx), \
+				     CONFIG_KERNEL_VM_BASE,		\
+				     CONFIG_KERNEL_VM_SIZE),		\
+		     "Extended region should not overlap virtual ram");	\
+	SYS_MEM_BLOCKS_DEFINE_STATIC_WITH_EXT_BUF(			\
+		region_##idx, XEN_PAGE_SIZE,				\
+		XEN_PFN_DOWN(DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx)),	\
+		(uint8_t *)DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx));
+
+DT_FOREACH_REG(XEN_HYP_NODE, EXT_BLOCK);
+
+#define EXT_REG_POINTER(node_id, idx)					\
+	{								\
+		.block = &region_##idx,					\
+		.phys_addr = DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),	\
+		.size = DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx),		\
+	},
+
+/*
+ * TODO: For now we allocate buffer structures for DT_NUM_REGS, when we
+ * need a range 1...DT_FOREACH_REG. Although region 0 is ignored, it still
+ * takes space for sys_block structures allocation. This requires optimization
+ */
+static struct extended_regions ext_blocks[DT_NUM_REGS(XEN_HYP_NODE)] = {
+	DT_FOREACH_REG(XEN_HYP_NODE, EXT_REG_POINTER)
+};
+
+static int addr_is_in_memblock(uint8_t *ptr, sys_mem_blocks_t *block)
+{
+	/* Compare ptr to buffer address to determine if it was allocated in this block */
+	return (ptr >= block->buffer) &&
+		(ptr < block->buffer + (block->info.num_blocks << block->info.blk_sz_shift));
+}
+
+bool xen_region_is_addr_extreg(void *ptr)
+{
+	for (unsigned int i = EXTENDED_REGIONS_START_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		if (addr_is_in_memblock(ptr, ext_blocks[i].block)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+void *xen_region_get_pages(size_t nr_pages)
+{
+	int i;
+	int ret = 0;
+	void *ptr = NULL;
+
+	/*
+	 * Locking ranges iteration, although each call sys_mem_blocks_alloc_contiguous
+	 * has it's own syncronisation, to avoid the situation of 2 parralel calls of
+	 * get_page and put_page on full buffers when put_page clears space in block 1
+	 * when get_page has already checked block 1. This will lead us to -ENOMEM while
+	 * block 1 space was freed already
+	 */
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_START_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		ret = sys_mem_blocks_alloc_contiguous(ext_blocks[i].block, nr_pages, &ptr);
+		if (ret == 0) {
+			break;
+		}
+	}
+
+	if (ret) {
+		LOG_ERR("xen_regions: unable to allocate pages, ret = %d", ret);
+		ptr = NULL;
+	}
+
+	k_mutex_unlock(&ext_regions_lock);
+
+	return ptr;
+}
+
+int xen_region_put_pages(void *ptr, size_t nr_pages)
+{
+	int i;
+	int ret = 0;
+	sys_mem_blocks_t *block = NULL;
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_START_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		if (addr_is_in_memblock(ptr, ext_blocks[i].block)) {
+			block = ext_blocks[i].block;
+			break;
+		}
+	}
+
+	if (!block) {
+		LOG_ERR("xen_regions: unable to find buffer, matching to addr %p", ptr);
+		ret = -EIO;
+		goto unlock;
+	}
+
+	ret = sys_mem_blocks_free_contiguous(block, ptr, nr_pages);
+	if (ret) {
+		LOG_ERR("xen_regions: unable to free pages, ret = %d", ret);
+		goto unlock;
+	}
+
+unlock:
+	k_mutex_unlock(&ext_regions_lock);
+	return ret;
+}
+
+int xen_region_map(void *ptr, size_t nr_pages)
+{
+	if (!xen_region_is_addr_extreg(ptr)) {
+		LOG_ERR("xen_regions: trying to map memory (%p) outside of any of regions", ptr);
+		return -EFAULT;
+	}
+	k_mem_map_phys_bare((uint8_t **)&ptr, (uintptr_t)ptr,
+						nr_pages * XEN_PAGE_SIZE,
+						K_MEM_CACHE_WB | K_MEM_PERM_RW | K_MEM_DIRECT_MAP);
+	return 0;
+}
+
+int xen_region_unmap(void *ptr, size_t nr_pages)
+{
+	if (!xen_region_is_addr_extreg(ptr)) {
+		LOG_ERR("xen_regions: trying to unmap memory (%p) outside of any of regions", ptr);
+		return -EFAULT;
+	}
+	k_mem_unmap_phys_bare(ptr, nr_pages * XEN_PAGE_SIZE);
+	return 0;
+}
diff --git a/include/zephyr/xen/regions.h b/include/zephyr/xen/regions.h
new file mode 100644
index 00000000000..14d128ec497
--- /dev/null
+++ b/include/zephyr/xen/regions.h
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2026 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Xen extended regions interface
+ *
+ * This file provides the interface for managing Xen extended regions,
+ * including allocation, freeing pages, and mapping/unmapping extended
+ * regions memory.
+ */
+
+#ifndef ZEPHYR_INCLUDE_XEN_REGIONS_H_
+#define ZEPHYR_INCLUDE_XEN_REGIONS_H_
+
+/**
+ * Checks if the provided address belongs to the Xen extended regions
+ *
+ * @param ptr the address/pointer to be checked
+ * @return true if the address is located within an extended region
+ * @return false if the address is outside all extended regions or if ptr is NULL
+ */
+bool xen_region_is_addr_extreg(void *ptr);
+
+/**
+ * Allocate pages from the extended regions
+ *
+ * @param nr_pages number of pages to be allocated
+ * @return pointer to the allocated pages
+ */
+void *xen_region_get_pages(size_t nr_pages);
+
+/**
+ * Free pages on extended regions
+ *
+ * @param ptr pointer to the pages, allocated by xen_region_get_pages call
+ * @param nr_pages number of pages that were allocated
+ * @return zero on success, non-zero on failure
+ */
+int xen_region_put_pages(void *ptr, size_t nr_pages);
+
+/**
+ * Map extended region memory
+ *
+ * @param ptr pointer to the pages, allocated by xen_region_get_pages call
+ * @param nr_pages number of pages that should be mapped
+ * @return zero on success, non-zero on failure
+ */
+int xen_region_map(void *ptr, size_t nr_pages);
+
+/**
+ * Unmap extended region memory
+ *
+ * @param ptr pointer to the pages, allocated by xen_region_get_pages call
+ * @param nr_pages number of pages that should be unmapped
+ * @return zero on success, non-zero on failure
+ */
+int xen_region_unmap(void *ptr, size_t nr_pages);
+
+#endif /* ZEPHYR_INCLUDE_XEN_REGIONS_H_ */
